pageserver/src/tenant/layer_map/historic_layer_coverage.rs

```diff
@@ -0,0 +1,432 @@
+/// we take an additional log(N) performance hit for queries, which will probably
+/// still be more critical.
+///
+/// See this for more on persistent and retroactive techniques:
+/// https://www.youtube.com/watch?v=WqCWghETNDc&t=581s
+pub struct BufferedHistoricLayerCoverage<Value> {
+    /// A persistent layer map that we rebuild when we need to retroactively update
+    historic_coverage: HistoricLayerCoverage<Value>,
+
+    /// We buffer insertion into the PersistentLayerMap to decrease the number of rebuilds.
+    ///
+    /// We implicitly assume that layers are identified by their key and lsn range.
+    /// A value of None means we want to delete this item.
+    buffer: BTreeMap<(u64, u64, i128, i128, bool), Option<Value>>,
+
+    /// All current layers. This is not used for search. Only to make rebuilds easier.
+    ///
+    /// We implicitly assume that layers are identified by their key and lsn range.
+    layers: BTreeMap<(u64, u64, i128, i128, bool), Value>,
+}
+
+impl<T: std::fmt::Debug> std::fmt::Debug for BufferedHistoricLayerCoverage<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RetroactiveLayerMap")
+            .field("buffer", &self.buffer)
+            .field("layers", &self.layers)
+            .finish()
+    }
+}
+
+impl<T: Clone> Default for BufferedHistoricLayerCoverage<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<Value: Clone> BufferedHistoricLayerCoverage<Value> {
+    pub fn new() -> Self {
+        Self {
+            historic_coverage: HistoricLayerCoverage::<Value>::new(),
+            buffer: BTreeMap::new(),
+            layers: BTreeMap::new(),
+        }
+    }
+
+    pub fn insert(
+        self: &mut Self,
+        key: Range<i128>,
+        lsn: Range<u64>,
+        value: Value,
+        is_image: bool,
+    ) {
+        self.buffer.insert(
+            (lsn.start, lsn.end, key.start, key.end, is_image),
+            Some(value.clone()),
+        );
+    }
+
+    pub fn remove(self: &mut Self, key: Range<i128>, lsn: Range<u64>, is_image: bool) {
+        self.buffer
+            .insert((lsn.start, lsn.end, key.start, key.end, is_image), None);
+    }
+
+    pub fn rebuild(self: &mut Self) {
+        // Find the first LSN that needs to be rebuilt
+        let rebuild_since: u64 = match self.buffer.iter().next() {
+            Some(((lsn_start, _, _, _, _), _)) => lsn_start.clone(),
+            None => return, // No need to rebuild if buffer is empty
+        };
+
+        // Apply buffered updates to self.layers
+        self.buffer.retain(|rect, layer| {
+            match layer {
+                Some(l) => {
+                    let existing = self.layers.insert(rect.clone(), l.clone());
+                    if existing.is_some() {
+                        // TODO this happened once. Investigate.
+                        // panic!("can't overwrite layer");
```

Comment:
with this assertion a lot of the on-demand download tests fail

Reply:
@problame Is this expected behavior from https://github.com/neondatabase/neon/issues/3369 ? Could the bug cause us to insert a layer that already exists?

Reply:
> Does this check get tripped by that, if the removal and insertion are both in the same batch of buffered updates?

Yes that's definitely a bug in my code :) I'll fix it and re-enable the check to see if it's the only bug

Reply:
Hm actually I'll remove the check instead. I put it there to validate my assumption that layers are uniquely identified by their bounds. But that assumption is true based on the fact that a layer contains **all** updates within its range.